src/pqueue.c

```diff
@@ -0,0 +1,294 @@
+ * elements allocated on the Stack (stack memory of function calls).
+ *
+ * @param[in] pqueue the Priority Queue to be destroyed
+ */
+
+void pqueue_destroy_free(PQueue *pqueue)
+{
+    array_destroy_free(pqueue->v);
+    pqueue->mem_free(pqueue);
+}
+
+/*
+ * Pushes the element in the pqueue
+ * @param[in] pqueue the priority queue in which the element is to be pushed
+ * @param[in] element the element which is needed to be pushed
+ *
+ * @return CC_OK if the element was successfully pushed, or CC_ERR_ALLOC
+ * if the memory allocation for the new element failed.
+ */
+
+enum cc_stat pqueue_push(PQueue *pqueue, void *element)
+{
+
+    int i = array_size(pqueue->v);
+    void *child, *parent;
+    enum cc_stat status = array_add(pqueue->v, element);
+    if(status == CC_ERR_ALLOC)
+        return status;
+    if(i == 0)
+        return CC_OK;
+    array_get_at(pqueue->v, i, &child);
+    array_get_at(pqueue->v, cc_parent(i), &parent);
+    while(i != 0 && pqueue->comparator(child, parent) > 0)
+    {
+        array_swap_at(pqueue->v, i, cc_parent(i));
+        i = cc_parent(i);
+        array_get_at(pqueue->v, i, &child);
+        array_get_at(pqueue->v, cc_parent(i), &parent);
+    }
+    return CC_OK;
+}
+
+/*
+ * Gets the most prioritized element from the queue without popping it
+ * @param[in] pqueue the PQueue structure of which the top element is needed
+ * @param[out] out pointer where the element is stored
+ *
+ * @return CC_OK if the element was found, or CC_ERR_VALUE_NOT_FOUND if the
+ * PQueue is empty.
+ */
+
+enum cc_stat pqueue_top(PQueue *pqueue, void **out)
+{
+    return array_get_at(pqueue->v, 0, out);
+}
+
+/*
+ * Removes the most prioritized element from the PQueue
+ * @param[in] pqueue the PQueue structure whose element is needed to be popped
+ * @param[out] out the pointer where the removed element will be stored
+ *
+ * return CC_OK if the element was popped successfully, or CC_ERR_OUT_OF_RANGE
+ * if pqueue was empty
+ */
+
+enum cc_stat pqueue_pop(PQueue *pqueue, void **out)
+{
+    void *tmp;
+    if(array_size(pqueue->v) == 0)
+        return CC_ERR_OUT_OF_RANGE;
+    array_swap_at(pqueue->v, 0, array_size(pqueue->v) - 1);
+    array_remove_last(pqueue->v, &tmp);
+    cc_heapify(pqueue, 0);
+    if(out)
+        *out = tmp;
+    return CC_OK;
+}
+
+/*
+ * A default comparator function for the ints.
+ * @param[in] two pointers a and b which are needed to be compared, they are
+ *     stored in the PQueue structure
+ * @return positive value if value at a is greater than at b, 0 if value at a
+ *     is equal to at of b, negative value if a is smaller than b.
+ */
+
+static int cc_default_comparator(const void *a, const void *b)
+{
+    return *((int *) a) - *((int *) b);
+}
+
+/*
+ * A function used to find the parent of the node number i.
+ * @param[in] a, the index of node in the heap
+ * @return a non-negative integer, which is the index of the parent of node i
+ */
+
+static size_t cc_parent(size_t a)
```

Comment:
A function seems a bit heavy for such a trivial calculation. A simple macro might be better choice. Same goes for the other two.